nemo_text_processing/g2p/modules.py

```diff
@@ -348,36 +373,82 @@ def __init__(
+                CMUdict-like entries. For example,
+                a dictionary file: scripts/tts_dataset_files/ipa_cmudict-0.7b_nv22.06.txt;
+                a dictionary object: {..., "Wire": [["ˈ", "w", "a", "ɪ", "ɚ"], ["ˈ", "w", "a", "ɪ", "ɹ"]], ...}.
+
+        Returns: a dict object (Dict[str, List[List[str]]]).
+        """
         if isinstance(phoneme_dict, str) or isinstance(phoneme_dict, pathlib.Path):
-            # load the dictionary file where there may exist a digit suffix after a word, which
+            # load the dictionary file where there may exist a digit suffix after a word, e.g. "Word(2)", which
             # represents the pronunciation variant of that word.
             phoneme_dict_obj = defaultdict(list)
             _alt_re = re.compile(r"\([0-9]+\)")
             with open(phoneme_dict, "r") as fdict:
                 for line in fdict:
-                    if len(line) and ('A' <= line[0] <= 'Z' or line[0] == "'"):
+                    # skip the empty lines
+                    if len(line) == 0:
+                        continue
+
+                    # Note that latin character pattern should be consistent with
+                    # nemo_text_processing.g2p.data.data_utils.LATIN_CHARS_ALL. It is advised to extend its character
+                    # coverage if adding the support of new languages.
+                    # TODO @xueyang: unify hardcoded range of characters with LATIN_CHARS_ALL to avoid duplicates.
+                    line = normalize_unicode_text(line)
+
+                    if (
+                        'A' <= line[0] <= 'Z'
+                        or 'a' <= line[0] <= 'z'
+                        or 'À' <= line[0] <= 'Ö'
+                        or 'Ø' <= line[0] <= 'ö'
+                        or 'ø' <= line[0] <= 'ÿ'
+                        or line[0] == "'"
+                    ):
                         parts = line.strip().split(maxsplit=1)
                         word = re.sub(_alt_re, "", parts[0])
                         prons = re.sub(r"\s+", "", parts[1])
                         phoneme_dict_obj[word].append(list(prons))
         else:
             # Load phoneme_dict as dictionary object
-            logging.info("Loading phoneme_dict as a Dict object.")
-            phoneme_dict_obj = phoneme_dict
+            logging.info("Loading phoneme_dict as a Dict object, and validating its entry format.")
+
+            phoneme_dict_obj = {}
+            for word, prons in phoneme_dict.items():
+                # validate dict entry format
+                assert isinstance(
+                    prons, list
+                ), f"Pronunciation type <{type(prons)}> is not supported. Please convert to <list>."
+
+                # normalize word with NFC form
+                word = normalize_unicode_text(word)
+
+                # normalize phonemes with NFC form
+                prons = [[normalize_unicode_text(p) for p in pron] for pron in prons]
```

Comment:
Do we expect to need to normalize the phonemes? Is it safe to assume that we can conflate two distinct IPA symbols that are the same after normalization?

Reply:
the motivation here is similar to graphemes. In IPA symbol sets, it usually combines a base token with its diacritics, like a nasalized `/ẽ/`. Without this normalization, `/ẽ/` would be encoded like `/e/ + / ̃/`